File: learning-servers/node/src/validate.js
export const validateSchema = {
  body: {
    type: "object",
    required: ["username", "email", "age", "password", "website", "country", "tags", "items"],
    additionalProperties: false,
    properties: {
      username: { type: "string", minLength: 3, maxLength: 30, pattern: "^[a-z0-9_]+$" },
      email: { type: "string", maxLength: 100, pattern: "^[^@\\s]+@[^@\\s]+\\.[^@\\s]+$" },
      age: { type: "integer", minimum: 13, maximum: 120 },
      password: { type: "string", minLength: 8, maxLength: 100, pattern: "^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9]).+$" },
      website: { type: "string", maxLength: 200, pattern: "^https?://" },
      country: { type: "string", enum: ["US", "CA", "GB", "DE", "FR", "JP", "AU", "BR", "IN", "CN"] },
      tags: { type: "array", minItems: 1, maxItems: 10, items: { type: "string", minLength: 1, maxLength: 20 } },
      items: {
        type: "array",
        minItems: 1,
        maxItems: 50,
        items: {
          type: "object",
          required: ["sku", "qty", "price"],
          additionalProperties: false,
          properties: {
            sku: { type: "string", pattern: "^[A-Z]{3}-[0-9]{3}$" },
            qty: { type: "integer", minimum: 1, maximum: 999 },
            price: { type: "number", minimum: 0, maximum: 100000 },
          },
        },
      },
    },
  },
};

export const validate = async () => ({ valid: true });

File: learning-servers/rust/src/validate.rs
use axum::extract::rejection::JsonRejection;
use axum::http::StatusCode;
use axum::response::IntoResponse;
use axum::Json;
use garde::Validate;
use serde::Deserialize;

#[derive(Deserialize, Validate)]
#[serde(deny_unknown_fields)]
pub struct Item {
    #[garde(pattern(r"^[A-Z]{3}-[0-9]{3}$"))]
    sku: String,
    #[garde(range(min = 1, max = 999))]
    qty: i64,
    #[garde(range(min = 0.0, max = 100_000.0))]
    price: f64,
}

#[derive(Deserialize, Validate)]
#[serde(deny_unknown_fields)]
pub struct Payload {
    #[garde(length(min = 3, max = 30), pattern(r"^[a-z0-9_]+$"))]
    username: String,
    #[garde(length(max = 100), pattern(r"^[^@ ]+@[^@ ]+\.[^@ ]+$"))]
    email: String,
    #[garde(range(min = 13, max = 120))]
    age: i64,
    #[garde(length(min = 8, max = 100), custom(password_complexity))]
    password: String,
    #[garde(length(max = 200), pattern(r"^https?://"))]
    website: String,
    #[garde(custom(country_valid))]
    country: String,
    #[garde(length(min = 1, max = 10), inner(length(min = 1, max = 20)))]
    tags: Vec<String>,
    #[garde(length(min = 1, max = 50), dive)]
    items: Vec<Item>,
}

fn password_complexity(v: &str, _: &()) -> garde::Result {
    let lower = v.chars().any(|c| c.is_ascii_lowercase());
    let upper = v.chars().any(|c| c.is_ascii_uppercase());
    let digit = v.chars().any(|c| c.is_ascii_digit());
    if lower && upper && digit {
        Ok(())
    } else {
        Err(garde::Error::new("must mix upper, lower, and digit"))
    }
}

fn country_valid(v: &str, _: &()) -> garde::Result {
    const COUNTRIES: [&str; 10] = [
        "US", "CA", "GB", "DE", "FR", "JP", "AU", "BR", "IN", "CN",
    ];
    if COUNTRIES.contains(&v) {
        Ok(())
    } else {
        Err(garde::Error::new("unknown country"))
    }
}

pub async fn validate(payload: Result<Json<Payload>, JsonRejection>) -> impl IntoResponse {
    let ok = matches!(payload, Ok(Json(p)) if p.validate().is_ok());
    let status = if ok { StatusCode::OK } else { StatusCode::BAD_REQUEST };
    (status, Json(serde_json::json!({ "valid": ok })))
}
